src/Interpreters/Cluster.cpp

```diff
@@ -668,7 +668,34 @@ std::unique_ptr<Cluster> Cluster::getClusterWithMultipleShards(const std::vector
     return std::unique_ptr<Cluster>{ new Cluster(SubclusterTag{}, *this, indices) };
 }
 
-Cluster::Cluster(Cluster::ReplicasAsShardsTag, const Cluster & from, const Settings & settings)
+namespace
+{
+
+void shuffleReplicas(auto & replicas, const Settings & settings)
+{
+    std::random_device rd;
+    std::mt19937 gen{rd()};
+
+    if (settings.prefer_localhost_replica)
+    {
+        // force for local replica to always be included
+        auto local_replica = std::find_if(replicas.begin(), replicas.end(), [](const auto & replica) { return replica.is_local; });
```

Comment:
Technically, it's possible to have > 1 local replicas.
Maybe we should handle it properly. (But I think it's a very rare case).